src/main/java/io/knowledgeassets/myskills/server/community/RecommendedCommunityRepositoryImpl.java

```diff
@@ -0,0 +1,57 @@
+package io.knowledgeassets.myskills.server.community;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+import org.neo4j.ogm.session.Session;
+import org.neo4j.ogm.session.SessionFactory;
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.data.neo4j.transaction.SessionFactoryUtils;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static java.util.Objects.requireNonNull;
+
+public class RecommendedCommunityRepositoryImpl implements RecommendedCommunityRepository {
+
+	private static final String QUERY = "MATCH (community:Community)-[:RELATES_TO]-(skill:Skill)-[r:RELATED_TO]-(user:User {id:{userId}}) " +
```

Comment:
I understand your concerns about maintainability. They are reasonable enough as the query does not seem to be easy-to-read one. Besides I have some concerns about performance of both ways (the multi-stage procedure you suggested and the query I wrote).

From my experience with relational databases it is worth trying to fetch data with a single query as these days relational databases are smart enough to optimize a query (if it is written properly) to execute it fast enough.
Though there are cases when a huge query should be split up into several small ones to increase its performance.

I cannot tell right now the extent Neo4j is smart enough to and whether it can optimize queries on its own.

There is another point regarding several small queries (multi-stage procedure)...The collections of objects of large size can be created when executing these queries. These collections will not live for a long time as they will be local ones but at some point of time they will consume the heap memory. That is why I try to hand over as much as I can to the database.

Reply:
Good points, I understand your arguments. 👍 After all, it's a trade-off. As I said, you can leave it like this at the moment. As soon as we touch this logic again we should think about a reasonable strategy how to ensure maintainability/readability and query performance at the same time.